src/main/scala/org/ergoplatform/nodeView/wallet/ErgoAddress.scala

```diff
@@ -0,0 +1,240 @@
+  }
+
+  override def equals(obj: Any): Boolean = obj match {
+    case p2sh: Pay2SHAddress => util.Arrays.equals(scriptHash, p2sh.scriptHash)
+    case _ => false
+  }
+
+  override def hashCode(): Int = Ints.fromByteArray(scriptHash.takeRight(4))
+
+  override def toString: String = encoder.toString(this)
+}
+
+object Pay2SHAddress {
+  def apply(pubkey: ProveDlog)(implicit encoder: ErgoAddressEncoder): Pay2SHAddress = apply(pubkey)
+
+  def apply(script: Value[SBoolean.type])(implicit encoder: ErgoAddressEncoder): Pay2SHAddress = {
+    val sb = ValueSerializer.serialize(script)
+    val sbh = ErgoAddressEncoder.hash192(sb)
+    new Pay2SHAddress(sbh)
+  }
+
+  val addressTypePrefix: Byte = 2: Byte
+}
+
+class Pay2SAddress(override val script: Value[SBoolean.type],
+                   val scriptBytes: Array[Byte])
+                  (implicit val encoder: ErgoAddressEncoder) extends ErgoAddress {
+  override val addressTypePrefix: Byte = Pay2SAddress.addressTypePrefix
+
+  override val contentBytes: Array[Byte] = scriptBytes
+
+  override def equals(obj: Any): Boolean = obj match {
+    case p2s: Pay2SAddress => util.Arrays.equals(scriptBytes, p2s.scriptBytes)
+    case _ => false
+  }
+
+  override def hashCode(): Int = Ints.fromByteArray(scriptBytes.takeRight(4))
+
+  override def toString: String = encoder.toString(this)
+}
+
+object Pay2SAddress {
+  def apply(script: Value[SBoolean.type])(implicit encoder: ErgoAddressEncoder): Pay2SAddress = {
+    val sb = ValueSerializer.serialize(script)
+    new Pay2SAddress(script, sb)
+  }
+
+  val addressTypePrefix: Byte = 3: Byte
+}
+
+
+case class ErgoAddressEncoder(settings: ErgoSettings) {
+
+  import ErgoAddressEncoder._
+
+  implicit private val ergoAddressEncoder: ErgoAddressEncoder = this
+
+  val ChecksumLength = 4
+
+  private val networkPrefix = settings.chainSettings.addressPrefix
+
+  def toString(address: ErgoAddress): String = {
+    val withNetworkByte = (networkPrefix + address.addressTypePrefix).toByte +: address.contentBytes
+
+    val checksum = hash256(withNetworkByte).take(ChecksumLength)
+    Base58.encode(withNetworkByte ++ checksum)
+  }
+
+  def fromString(addrStr: String): Try[ErgoAddress] = Base58.decode(addrStr).flatMap { bytes =>
+    Try {
+      val headByte = bytes.head
+      require(headByte >= networkPrefix)
+      val addressType = (headByte - networkPrefix).toByte
+      val (withoutChecksum, checksum) = bytes.splitAt(bytes.length - ChecksumLength)
+
+      if (!util.Arrays.equals(hash256(withoutChecksum).take(ChecksumLength), checksum)) {
+        throw new Exception(s"Checksum check fails for $addrStr")
+      }
+
+      val bs = withoutChecksum.tail
+
+      addressType match {
+        case P2PKAddress.addressTypePrefix =>
+          val buf = ByteBuffer.wrap(bs)
+          val r = new ByteBufferReader(buf).mark()
+          val ge = DataSerializer.deserialize[SGroupElement.type](SGroupElement, r)
+          val pd = ProveDlog(ge)
```

Comment:
@kushti After using `ValueSerializer` for P2PK serialization (per my suggestion) we need to deserialize ProveDlog here with `val pd = ValueSerializer.deserialize(r).asInstanceOf[ProveDlog]` since `ProveDlogSerializer` prefixes with it's op code byte. If we want to remove this extra byte we should switch back to serializing the underlying group element. Sorry I missed that in my initial suggestion.

Reply:
@greenhat  fixed!